packages/yew-router-macro/src/routable_derive.rs

```diff
@@ -22,225 +26,457 @@ impl Parse for Routable {
+            BindType::String => quote! { args.pop_str() },
+        }
+    }
+    fn encoder(&self, arg: &Ident) -> TokenStream {
+        let hidden = hidden_module();
+        match self {
+            BindType::Rest => quote! { args.store_route(#hidden::Routable::to_route(#arg)); },
+            BindType::String => quote! { args.push_str(#arg); },
+        }
+    }
+}
+
+struct ConstMuncher {
+    muncher: Option<TokenStream>,
+    binds: Vec<(usize, BindType)>,
+}
+
+fn generate_path_muncher(fields: &[ParsedField], at: &ParsedAt) -> syn::Result<ConstMuncher> {
+    let hidden = hidden_module();
+    let mut binds = Vec::new();
+    for name in &at.names {
+        if let Some(field) = fields.iter().find(|field| &field.name == name) {
+            binds.push((field.index, BindType::String));
+        } else {
+            return Err(syn::Error::new(
+                at.span,
+                format_args!("No field named {:?}", name),
+            ));
+        }
+    }
+    let parts: TokenStream = at
+        .parts
+        .iter()
+        .map(|part| match part {
+            PathPart::Match(s) => quote! { #hidden::PathPart::Match(#s), },
+            PathPart::ExtractOne => quote! { #hidden::PathPart::ExtractOne, },
+            PathPart::ExtractAll => quote! { #hidden::PathPart::ExtractAll, },
         })
+        .collect();
+    Ok(ConstMuncher {
+        muncher: Some(quote! {
+            #hidden::PathSegmentMuncher {
+                parts: &[#parts]
+            }
+        }),
+        binds,
+    })
+}
+
+fn generate_arg_muncher(muncher: &Ident, fields: &[ParsedField]) -> syn::Result<ConstMuncher> {
+    let hidden = hidden_module();
+    let mut binds = Vec::new();
+    let mut names = Vec::new();
+    for field in fields {
+        let name = match &field.meta {
+            Some(syn::Meta::Path(_)) => field.name.clone(),
+            Some(syn::Meta::NameValue(syn::MetaNameValue {
+                lit: syn::Lit::Str(name),
+                ..
+            })) => name.value(),
+            _ => {
+                return Err(syn::Error::new(
+                    field.meta.span(),
+                    "Invalid argument binding",
+                ))
+            }
+        };
+        names.push(name);
+        binds.push((field.index, BindType::String));
     }
+    Ok(ConstMuncher {
+        muncher: Some(quote! {
+            #hidden::#muncher {
+                names: &[#(#names,)*]
+            }
+        }),
+        binds,
+    })
 }
 
-fn parse_variants_attributes(
-    variants: &Punctuated<Variant, syn::token::Comma>,
-) -> syn::Result<(Option<Ident>, Vec<LitStr>)> {
-    let mut not_founds = vec![];
-    let mut ats: Vec<LitStr> = vec![];
+fn generate_rest_glue(fields: &[ParsedField]) -> syn::Result<ConstMuncher> {
+    let binds = match fields {
+        [field] => vec![(field.index, BindType::Rest)],
+        _ => {
+            return Err(syn::Error::new(
+                Span::call_site(),
```

Comment:
Let's use span of all the `rest` bindings instead of `call_site`

Reply:
`Span::join` is nightly-only. I'll use the span of the second `rest` binding for now.

Reply:
You can collect the tokens in a `TokenStream` and use `Span::new_spanned` with that stream for effectively the same effect

Reply:
👍